### PR TITLE
python@3.8 3.8.0 (new formula)

### PR DESCRIPTION
**Pull request by bayandin**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

This PR adds `python@3.8` formula ~~and migrates 10 most popular depending on `python` formulae (but `glib`)~~

### CODE REVIEW

**Comment by Bo98** [comment]
Hmm might need to be careful and think about dependency trees here.

If a Python 3.7 formula depends on a Python 3.8 formula, there's gonna be problems. I'm not sure how many of that there is though.

One thing for sure is Cython. I reckon we're gonna have to have two Cython formulae - one for Python 3.7 and one for Python 3.8.

**Comment by Bo98** [comment]
Actually a lot less depending on Cython than I thought. We'll need to migrate that and their dependents together, either now or later on.

**Comment by bayandin** [comment]
Here it is, the list of formulae depend on `cython` as a formula or as a resource:
```
adios2
apache-arrow
apache-arrow-glib
arcade-learning-environment
astrometry-net
aubio
boost-python3
breezy
btfs
caffe
cassandra
cassandra@2.2
csound
dvc
gdal
gmt
gmt@5
graph-tool
liblas
libtorrent-rasterbar
mapnik
mapserver
mpi4py
numpy
ola
opencv
opencv@3
openimageio
osm2pgrouting
pdal
pgrouting
points2grid
postgis
scipy
simple-tiles
siril
urh
virtualpg
visp
wxpython

numpy@1.16
pytouhou
urh
vapoursynth
gnuradio
gr-osmosdr
opencv@2
mpv
mvtools
vapoursynth-imwri
vapoursynth-ocr
vapoursynth-sub
```

**Comment by Bo98** [comment]
As a resource doesn't really matter since updating the Cython formula won't mandate updating those that use Cython as a resource. As a formula it does matter. And it needs to be done recursively: `brew uses --include-build --include-test --recursive cython`.

You then need to watch out if anything on that list depends on another Python formula that doesn't itself depend on Cython - that too will need to be updated. Not sure of a nice way to get such list.

**Comment by bayandin** [comment]
This PR will land `python@3.8` formula only (because bumping the first 10 formulae from the list has eaten all the space on CI again)

File: Formula/python@3.8.rb [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+      Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to
+      `python3`, `python3-config`, `pip3` etc., respectively, have been installed into
+        #{opt_libexec}/bin
+      If you need Homebrew's Python 2.7 run
+        brew install python@2
+      You can install Python packages with
+        pip3 install <package>
+      They will install into the site-package directory
+        #{HOMEBREW_PREFIX/"lib/python#{xy}/site-packages"}
+      See: https://docs.brew.sh/Homebrew-and-Python
```
Comment (bayandin) [inline Formula/python@3.8.rb]: It seems I need to tune these lines a bit

Comment (bayandin) [reply Formula/python@3.8.rb]: Removed for the migration period

**Comment by michaelblyons** [comment]
Can an owner add the [python] label to this PR?

**Comment by SMillerDev** [comment]
It doesn't really change anything but here you go

**Comment by OJFord** [comment]
Is there anything still needed here? I understand it's just step 1 in a larger effort to get to `python` being at 3.8, but this PR alone is helpful for people that want to use it directly, keeping old `python` around for any dependencies.

File: Formula/python@3.8.rb [hunk 2]
```diff
@@ -0,0 +1,287 @@
+class PythonAT38 < Formula
+  desc "Interpreted, interactive, object-oriented programming language"
+  homepage "https://www.python.org/"
+  url "https://www.python.org/ftp/python/3.8.0/Python-3.8.0.tar.xz"
+  sha256 "b356244e13fb5491da890b35b13b2118c3122977c2cd825e3eb6e7d462030d84"
+  # head "https://github.com/python/cpython.git" # Commented out for migration period (3.7 -> 3.8)
```
Comment (iMichka) [inline Formula/python@3.8.rb]: Please just remove comment and the head link. I think it's better to not have too much commented out code. We can re-add whatever is needed later.

File: Formula/python@3.8.rb [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+  pour_bottle? do
+    reason <<~EOS
+      The bottle needs the Apple Command Line Tools to be installed.
+        You can install them, if desired, with:
+          xcode-select --install
+    EOS
+    satisfy { MacOS::CLT.installed? }
+  end
+
+  keg_only :versioned_formula # Marked as keg_only for migration period (3.7 -> 3.8)
```
Comment (iMichka) [inline Formula/python@3.8.rb]: Please remove the comment, as above.

**Comment by iMichka** [comment]
I think we should add the caveats back (and adapt the paths in the caveat). Except that, everything looks good for me.

**Comment by bayandin** [comment]
Thanks @iMichka, added caveats back

File: Formula/python@3.8.rb [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+
+      Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to
+      `python3`, `python3-config`, `pip3` etc., respectively, have been installed into
+        #{opt_libexec}/bin
+
+      If you need Homebrew's Python 2.7 run
+        brew install python@2
+
+      You can install Python packages with
+        pip3 install <package>
```
Comment (iMichka) [inline Formula/python@3.8.rb]: As this is keg_only, I guess that the pip3 path should also be something like `#{HOMEBREW_PREFIX/"bin/pip3"}` ?

**Review by iMichka — APPROVED** [review]

**Comment by iMichka** [comment]
Let's merge this as-is once the build is done and green :)

**Comment by iMichka** [comment]
I'll open another PR to fix the caveats, I noticed that some information in there is wrong.

**Comment by iMichka** [comment]
Here is the PR to update the caveat: #47984

**Comment by bayandin** [comment]
Thanks @iMichka!

**Review by chenrui333 — COMMENTED** [review]
Good to see this move forward!!

**Pull request closed without merging**